File: Python/0107-e-level-order-traversal-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
BFS
"""
class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        queue = collections.deque([(root, 0)])
        res = []
        while queue:
            node, level = queue.popleft()
            if not node:
                continue
            if len(res) < level + 1:
                res.insert(0, [])
            res[-(level + 1)].append(node.val)
            queue.append((node.left, level + 1))
            queue.append((node.right, level + 1))
        return res

File: leetcode/C++/0107-e-level-order-traversal-ii.cpp
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode *root) {
        vector<vector<int>> result;
        traverse(root, 1, result);
        std::reverse(result.begin(), result.end());
        return result;
    }
private:
    void traverse(TreeNode *root, size_t level, vector<vector<int>> &result) {
        if (!root)
            return;
        
        if (level > result.size())
            result.push_back(vector<int>());
        
        result[level-1].push_back(root->val);
        traverse(root->left, level+1, result);
        traverse(root->right, level+1, result);
    }
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        queue<TreeNode *> q;

        if (root == nullptr)
            return result;
        else
            q.push(root);
        
        while (!q.empty()) {
            vector<int> level;
            size_t size = q.size();
            for (size_t i = 0; i < size; i++) {
                TreeNode *node = q.front();
                q.pop();
                level.push_back(node->val);
                if (node->left != nullptr)
                    q.push(node->left);
                if (node->right != nullptr)
                    q.push(node->right);
            }
            result.push_back(level);
        }
        reverse(result.begin(), result.end());
        return result;
    }
};